File: 90_Subsets_II/Solution.py
class Solution:
    def subsetsWithDup(self, nums):
        if None == nums or [] == nums:
            return nums

        n = len(nums)
        nums = sorted(nums)
        i = 0
        result = [[]]
        while i < n:
            c = 1
            while i< n and i+1<n and nums[i] == nums[i+1]:
                c += 1
                i += 1
            rl = len(result)
            for w in range(rl):
                for j in range(1,c+1):
                    tt = list(result[w])
                    for k in range(j):
                        tt.append(nums[i])
                    result.append(tt)
            
            i += 1

        return result
s = Solution()
print(s.subsetsWithDup([1,2,2]))

File: 90_Subsets_II/Solution.java
import java.util.*;

public class Solution {
    public List<List<Integer>> subsetsWithDup(int[] num) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> empty = new ArrayList<Integer>();
        result.add(empty);
        Arrays.sort(num);
        
        for (int i = 0; i < num.length; i++) {
            int dupCount = 0;
            while( ((i+1) < num.length) && num[i+1] == num[i]) {
                dupCount++;
                i++;
            }
            int prevNum = result.size();
            for (int j = 0; j < prevNum; j++) {
                List<Integer> element = new ArrayList<Integer>(result.get(j));
                for (int t = 0; t <= dupCount; t++) {
                    element.add(num[i]);
                    result.add(new ArrayList<Integer>(element));
                }
            }
        }

        return result;
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        int[] num = new int[]{1,2,2};
        List l = s.subsetsWithDup(num);
        l.forEach(list->{
            System.out.println(list.toString());
        });
    }
}